src/apprt/gtk/close_dialog.zig

```diff
@@ -0,0 +1,173 @@
+        const super = self.as(Parent);
+
+        // If we don't have a possible window to ask the user,
+        // in most situations (e.g. when a split isn't attached to a window)
+        // we should just close unconditionally.
+        const dialog_window = target.dialogWindow() orelse {
+            target.close();
+            self.as(gtk.Widget).unref();
+            return;
+        };
+
+        self.private().target = target;
+        super.setHeading(target.title());
+        super.setBody(target.body());
+
+        super.choose(
+            dialog_window.as(gtk.Widget),
+            null,
+            closeCallback,
+            null,
+        );
+    }
+
+    fn init(self: *CloseDialog, _: *Class) callconv(.C) void {
+        const dialog = self.as(Parent);
+        dialog.addResponse("cancel", "Cancel");
+        dialog.setCloseResponse("cancel");
+        dialog.setDefaultResponse("cancel");
+
+        dialog.addResponse("close", "Close");
+        dialog.setResponseAppearance("close", .destructive);
+    }
+
+    fn closeCallback(
+        source_object: ?*gobject.Object,
+        result: *gio.AsyncResult,
+        _: ?*anyopaque,
+    ) callconv(.C) void {
+        const self = gobject.ext.cast(CloseDialog, source_object.?).?;
+        const resp = self.as(Parent).chooseFinish(result);
+
+        if (std.mem.orderZ(u8, resp, "close") == .eq) {
+            self.private().target.close();
+        }
+    }
+
+    // More GObject boilerplate
+    pub fn as(self: *CloseDialog, comptime T: type) *T {
+        return gobject.ext.as(T, self);
+    }
+
+    fn private(self: *CloseDialog) *Private {
+        return gobject.ext.impl_helpers.getPrivate(self, Private, Private.offset);
+    }
+
+    pub const Class = extern struct {
+        parent: Parent.Class,
+
+        pub const Instance = CloseDialog;
+    };
+
+    /// The target of this close dialog.
+    ///
+    /// This is here so that we can consolidate all logic related to
+    /// prompting the user and closing windows/tabs/surfaces/etc.
+    /// together into one struct that is the sole source of truth.
+    pub const Target = union(enum) {
+        app: *App,
+        window: *Window,
+        tab: *Tab,
+        surface: *Surface,
+
+        pub fn title(self: Target) [:0]const u8 {
+            return switch (self) {
+                .app => "Quit Ghostty?",
+                .window => "Close Window?",
+                .tab => "Close Tab?",
+                .surface => "Close Surface?",
+            };
+        }
+
+        pub fn body(self: Target) [:0]const u8 {
+            return switch (self) {
+                .app => "All terminal sessions will be terminated.",
+                .window => "All terminal sessions in this window will be terminated.",
+                .tab => "All terminal sessions in this tab will be terminated.",
+                .surface => "The currently running process in this surface will be terminated.",
```

Comment:
Do we currently know what the running process is? See the GNOME Console pictures above.

Reply:
Not yet. That’s some core work that needs to be done. It’ll require some upstream changes in xev too. We need this for many planned features I want to build…